airbyte-cdk/python/airbyte_cdk/connector_builder/message_grouper.py

```diff
@@ -300,6 +308,17 @@ def _read_stream(
         # iterate over the generated messages. if next raise an exception, catch it and yield it as an AirbyteLogMessage
         try:
             yield from AirbyteEntrypoint(source).read(source.spec(self.logger), config, configured_catalog, state)
+        except AirbyteTracedException as traced_exception:
+            # Look for this message which indicates that it is the "final exception" raised by AbstractSource.
+            # If it matches, don't yield this as we don't need to show this in the Builder.
+            # This is somewhat brittle as it relies on the message string, but if they drift then the worst case
+            # is that this message will be shown in the Builder.
+            if (
+                traced_exception.message is not None
+                and "During the sync, the following streams did not sync successfully" in traced_exception.message
```

Comment:
❓ Can we modify `AbstractSource` to indicate its a "Final Exception"?

Reply:
@bnchrch Do you have a specific approach in mind? We could set `internal_message` to something for example to indicate this but this would still rely on comparing strings in that approach.

I'm wary to raise a different exception in abstract_source [here](https://github.com/airbytehq/airbyte/blob/d1d6e2333adcdd7435a9baf89adcf936e5ec7bba/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py#L177) because I'm not sure what relies on this being an AirbyteTracedException and I don't feel familiar enough with that fairly central CDK code to know what is safe to change

Reply:
I would also add that any raised AirbyteTracedException is sort of already implied to be the final exception. Within `abstract_source.py`, https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py#L171-L177 , during a read of a stream always try to catch any form of Exception or AirbyteTracedException thrown by a stream and wrap it into an emitted trace message. We only formally raise the "final" AirbyteTracedException because we need to end the process with a non-zero error code.

Granted the small gap is if we get an exception outside of that try/except block. But from my perspective, us trying to adjust the protocol or create new abstractions to capture what is effectively a work around feels like we're trying to over engineer a bit. The ideal design is that the platform can identify that merely receiving any AirbyteTracedException is enough to know that the sync was unsuccessful. And then we could get rid of this work around.

So TLDR: I think how we do it here is fine.